### ISSUE TITLE
[Design] Should the `PublishEvents` RPC block when the queue is full?

### ISSUE THREAD

**Issue by cmacknz** [elastic/elastic-agent-shipper#81]
The shipper `PublishEvents` [documentation](https://github.com/elastic/elastic-agent-shipper-client/blob/af78e67e186cdd291199400cba9f7de20f962e17/api/shipper.proto#L13) currently specifies that it should block until processing is completed and events have been enqueued:

```protobuf
service Producer {
 // Publishes a list of events via the Elastic agent shipper.
 // Blocks until all processing steps complete and data is written to the queue.
 //
 // If the queue could not accept some events from the request, this returns a successful response
 // containing a count of events that were accepted by the queue.
 // The client is expected to retry sending the rest of the events in a separate request.
 // 
 // The client is also expected to have some kind of backoff strategy
 //	in case of a reply with an accepted count < the amount of sent events.
 rpc PublishEvents(messages.PublishRequest) returns (messages.PublishReply)
```

The initial gRPC server implementation however does not implement this behaviour: https://github.com/elastic/elastic-agent-shipper/pull/76#discussion_r920137777.

The scope of this issue is to decide whether the `PublishEvents` RPC should block when the queue is full, and then update either the RPC documentation or the server implementation based on the outcome.

There were two example scenarios provided in the original discussion:

## Example 1. The queue filled up during the request (logic already in place)
1. The client requests to publish 50 events
2. The queue can accept only 5 and we get "queue is full" error back from publishing the 6th
3. The server replies to the client with the current AcceptedCount and AcceptedIndex values partially accepting the event batch

## Example 2. The queue is full from the beginning (needs to be implemented)
1. The client requests to publish 50 events
2. When publishing the first event, the queue returns "queue is full" error and cannot accept even a single event
3. The server blocks the client when calling the underlying queue's `Publish` call.
4. Once at least one event is accepted by the queue this switches to the behaviour in Example 

The behaviour proposed in the examples above are that `PublishEvents` would block until at least one event is accepted, and then return with the accepted count indicating how many events were accepted. The alternative would be blocking until all events are accepted.

**Comment by cmacknz** [comment]
Reproducing @rdner's original concerns with having the `PublishEvents` call block:

1. For Example 2 it's the lack of observability. From the client's perspective, this will look like just a slow server. The client would not know about the reason why the request is so slow.

2. A lot of clients will compete for the queue to become empty and this solution does not really solve anything regarding this point. We could, of course, implement a very sophisticated ordering system that will do "first come – first served" for the queue publishing requests but it will complicate things and might make certain requests extremely slow.

**Comment by cmacknz** [comment]
I personally prefer `PublishEvents` as a blocking call. I think it simplifies the client implementation in that they do not need to worry about retry strategies, they can simply call `PublishEvents` and wait for it to finish. 

It avoids having to deal with issues where a client has chosen a retry strategy poorly, in the worst case pointlessly consuming CPU cycles busy-waiting on the queue or introducing unnecessary latency with a too long retry loop. 

It also provides a natural and obvious back pressure signal where the input client should simply stop sending us new data.

The shipper currently integrates with the memory queue producer: https://github.com/elastic/elastic-agent-shipper/blob/cf6513d64e2e1343d9a0c7f4b29f3b333a455fc9/queue/queue.go#L57-L64

Looking at the beats' implementation of `Publish` for the memory and disk queues it is actually blocking by default. If we wanted non-blocking we would need to switch to `TryPublish` instead in the queue wrapper.

1. Beats' memory queue `publish` implementation: https://github.com/elastic/beats/blob/e6db9a53f2fff60b350153b03407bc38b21bf0b1/libbeat/publisher/queue/memqueue/produce.go#L128-L136
2. Beats' disk queue `publish` implementation: https://github.com/elastic/beats/blob/e6db9a53f2fff60b350153b03407bc38b21bf0b1/libbeat/publisher/queue/diskqueue/producer.go#L52-L62

With regards to @rdner's concerns above:

> For Example 2 it's the lack of observability. From the client's perspective, this will look like just a slow server. The client would not know about the reason why the request is so slow.

When a client learns that the queue is full, it only has one choice: wait some period of time and retry again. If we return a `QueueIsFull` error we delegate the choice of retry policy to the client, if we block we provide the optimal retry policy: immediately succeeding when space is free.

I am not worried about an input's ability to observe the queue state, but the agent user or operator's ability to observe the queue state. This we can provide through monitoring of the shipper itself.

> A lot of clients will compete for the queue to become empty and this solution does not really solve anything regarding this point. We could, of course, implement a very sophisticated ordering system that will do "first come – first served" for the queue publishing requests but it will complicate things and might make certain requests extremely slow.

Correct, we can't do anything about this without a more complex design. Both a retry-backoff based polling approach and a blocking approach have the same behaviour here: the client that "wins" access to the queue is non-deterministic right now.

**Comment by cmacknz** [comment]
@faec or @leehinman do you have any thoughts to add here?

**Comment by faec** [comment]
I think I prefer a fully blocking call if it's practical. Having RPCs routinely block for such a potentially long time makes me nervous abstractly, but since it's over local transport in our case I think it's ok. As far as concern (1), I agree this is possibly a false freedom for the client, since the knowledge that the queue is full shouldn't change the input's behavior (accumulate input data if it can, drop it if it must). For (2), when multiple clients are inserting into the queue in current implementations, having all of them block while waiting on progress should ensure eventual progress for each individual client (effectively they'll all be targeting the same channel, so progress should be as fair as the goroutine scheduler is).

One thing I'm uncertain of with blocking on everything is how we handle shutdown, since that's the only context where we'd practically need to know about partial progress in a long-running call. I believe our current expectation is that inputs will try to shut down before the shipper itself, so if it's technically feasible it would be nice to have some clean way to close off an active request and fetch the partial progress (this would be less relevant with non-blocking or blocking-on-first-event since that automatically gives more regular updates).

**Comment by leehinman** [comment]
I'm in favor of a blocking call, and the client should use a deadline on the gPRC call so they don't wait forever.  If the client wants to do a more complex retry they can based on the gRPC call termination with `DEADLINE_EXCEEDED` or they could just try again.

In the server we should probably be checking to see  if the context is cancelled as well so we can stop using resources when the client has given up and we can't report back to them.

**Comment by cmacknz** [comment]
Closing in favor of the implementation issue: https://github.com/elastic/elastic-agent-shipper/issues/84
